### PR TITLE
GODRIVER-1878 add unified-test-format/valid-fail tests

### PR DESCRIPTION
**Pull request by iwysiu**
Due to the changes I made for the atlas maintenance tests, I didn't actually have to make any continuous integration changes for this.

### CODE REVIEW

File: mongo/integration/unified/unified_spec_runner.go [hunk 1]
```diff
@@ -114,22 +126,42 @@ func RunTestFile(t *testing.T, filepath string, opts ...*Options) {
 			CreateClient(false)
 
 		mt.RunOpts(testCase.Description, mtOpts, func(mt *mtest.T) {
-			if err := testCase.Run(mt); err != nil {
-				mt.Fatal(err)
+			defer func() {
+				// catch panics from looking up elements and fail if it's unexpected
+				if r := recover(); r != nil {
```
Comment (iwysiu) [inline mongo/integration/unified/unified_spec_runner.go:131]: We use a lot of panicking raw lookup/casts and I can alternatively go through and change all of those to use _OK methods, but that's a considerable change

File: mongo/integration/unified/unified_spec_runner.go [hunk 2]
```diff
@@ -114,22 +126,42 @@ func RunTestFile(t *testing.T, filepath string, opts ...*Options) {
...
+				return
 			}
+			mt.Fatalf("Expected test to error, got nil")
 		})
 	}
 }
 
 // ParseTestFile create an array of TestCases from the testJSON json blob
-func ParseTestFile(t *testing.T, testJSON []byte, opts ...*Options) ([]mtest.RunOnBlock, []*TestCase) {
+func ParseTestFile(testJSON []byte, opts ...*Options) ([]mtest.RunOnBlock, []*TestCase, error) {
```
Comment (iwysiu) [inline mongo/integration/unified/unified_spec_runner.go:153]: I was trying to main the function signatures as much as possible, but there's no other way to get the error back from this.

File: mongo/integration/unified/entity.go [hunk 3]
```diff
@@ -368,6 +368,9 @@ func (em *EntityMap) addClientEntity(ctx context.Context, entityOptions *entityO
 	var client *clientEntity
 
 	for _, eventsAsEntity := range entityOptions.StoreEventsAsEntities {
+		if entityOptions.ID == eventsAsEntity.EventListID {
+			return fmt.Errorf("entity with ID %q already exists", entityOptions.ID)
+		}
```
Comment (benjirewis) [inline mongo/integration/unified/entity.go:373]: Was this change necessary for the `valid-fail` tests?

File: mongo/integration/unified/unified_spec_runner.go (hunk 1 above)
Comment (benjirewis) [reply mongo/integration/unified/unified_spec_runner.go:131]: Hmm seems out of scope for this ticket to me, and `recover`ing the `panic` seems acceptable.

File: mongo/integration/unified/unified_spec_runner.go (hunk 2 above)
Comment (benjirewis) [reply mongo/integration/unified/unified_spec_runner.go:153]: When would `valid/fail` expect an error from parsing? Wouldn't that imply that it's "invalid"?

**Review by kevinAlbs — COMMENTED** [review]
IIRC this is required for GODRIVER-1844? If so, can you update the description to note why.

> Due to the changes I made for the atlas maintenance tests, I didn't actually have to make any continuous integration changes for this.

Nice, `verifyOperationError` already handles the case where `ExpectedError` is nil and asserts that the operation did not have an error.

**Review by benjirewis — COMMENTED** [review]
LGTM with naming nit.

File: mongo/integration/unified/unified_spec_runner.go [hunk 4]
```diff
@@ -88,20 +89,26 @@ type TestFile struct {
 }
 
 // runTestDirectory runs the files in the given directory, which must be in the unifed spec format
-func runTestDirectory(t *testing.T, directoryPath string) {
+func runTestDirectory(t *testing.T, directoryPath string, pass bool) {
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply mongo/integration/unified/unified_spec_runner.go]: done

File: mongo/integration/unified/unified_spec_runner.go [hunk 5]
```diff
@@ -88,20 +89,26 @@ type TestFile struct {
 }
 
 // runTestDirectory runs the files in the given directory, which must be in the unifed spec format
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply mongo/integration/unified/unified_spec_runner.go]: done

**Review by benjirewis — APPROVED** [review]

File: mongo/integration/unified/unified_spec_runner.go [hunk 6]
```diff
@@ -87,21 +88,28 @@ type TestFile struct {
...
 		t.Run(filename, func(t *testing.T) {
-			RunTestFile(t, path.Join(directoryPath, filename))
+			runTestFile(t, path.Join(directoryPath, filename), expectValidFail)
 		})
 	}
 }
 
-// RunTestFile runs the tests in the given file, which must be in the unifed spec format
+// RunTestFile runs the tests in the given file, which must be in the unified spec format
 func RunTestFile(t *testing.T, filepath string, opts ...*Options) {
```
Comment (matthewdale) [inline mongo/integration/unified/unified_spec_runner.go:102]: Is there any reason to keep this `RunTestFile` function around? It doesn't seem to be used anywhere now that it's been replaced with `runTestFile`.

**Review by matthewdale — APPROVED** [review]
👍

**Review by kevinAlbs — APPROVED** [review]
LGTM!
